Storefront v2

janus-reith opened the pull request:
Resolves #661
Impact: **breaking**
Type: **refactor**

## Issue
TBD
Refactor to Hooks, Context instead of Mobx, i18n, latest nextjs + MUI and much more.

## Solution
TBD


## Testing
TBD

Commit:
Remove client side use of logger as its dependencies are not transpiled

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove useless/soon to be outdated tests

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove meaningless component readme's

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Update Material UI, styled-components, nextjs and react to latest, remove babel

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Update Material UI components

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove useless/soon to be outdated tests #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add Progressive Image from component library and alter ref, as this is breaking in styled-components | Todo: use by custom, better progressive image imlementation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add aliases to next for global imports without babel

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove use of decorators

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove unnecessary initialprops for viewport

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add new Apollo Implementation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add new Apollo Implementation #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove unused tracking

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Use new utiltiy functions for static data fetching

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Keep withShop for compatiblity, but instead of fetching pass  shop from props as getStaticProps provides it

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove useless withShop down in the tree

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove useless withShop down in the tree #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Comment out tracking for now, keep reference for later

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
import classic Apollo Wrappers from new lib, remove use of  nested withApollo within codebase

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Use Apollo per page

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Update fetch implementation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Update node version

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Import withApollo from react-components as this reexports it with with Query component compatibility

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
fix withsHop compatibility layer to use pageprops properly

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Replace next getConfig pattern with env

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
remove console logs

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remov withShop per page

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Implement new static catalog product fetch and use native routing to get slug

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
move out of src folder

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
move out of src folder #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Move deprecated static dir to public

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove usage of next-routes, use dynamic routes

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove custom server

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
dont omit env vars during build

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
set envalid strict to false

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
use api routes for existing express routes

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add static navigation and tag fetch

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Turn pdp page to functional component and show loader for fallback rendering(when a static page is requested the first time)

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
lintfix

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add new Context Stores

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add hooks to consume context

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add translations util

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add swr for token fetching

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add global Context providers

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove obsolete conainers

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove mobx usage

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add Context Providers for Tags and Navigation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove mobx usage - Fix

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
remove unused

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
remove unused use of session storage

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
useviewer query

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Wrap both Apollo Providers to support both hooks and query components

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Use Layout per page as _app is not wrapped by apollo

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add cart hook and wrap in hoc

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
remove pushRoute, use Route

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
adjust import

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
adjust imports

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
adjust imports

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Implement addressbook and payments hook, use in hoc

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Move shop hook

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Move to functional and seperate properly instead of rendering  by route

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Rename gql

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add remaining statc props

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add remaining statc props #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
lintfix

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
fix token checkout

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Wrap stripe provider on checkout, add static path fallback on checkout complete

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
use new graphql request wrapper function

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add cart reconcilation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Fix checkout route for autneticated users

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Fix proptypr

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Move pages into language subpath

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
adjust imports

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Implement i18n

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add locale paths for each page

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove logs

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Fix proptype

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Implement order/orders as hook, use in withOrder HOC

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Pass product url with locale

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Add margin

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
styling

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove duplicate import

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Fetch tag data statically per page

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add wrapper for useTranslation

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add example translations

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
add example translations #2

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Use Mui Button

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove rui theme

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove unsued click handler

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Reimplement Profil menu

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Wrap all context store providers in one HOC

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Fork CatalogGridItem t use dynamic pat routing

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
lintfix

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
only fetch all tags on pages using breadcrumbs - Todo: Get rif of fetchAllTags altogether

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Change refresh endpoint to return a JSON including hte new token instead of redirecting

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Silent reffresh: Introduce request queuing to the errorLink - Enqueue all requests that failed as unauthorized and wait for token refresh

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Applly linting rules

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Use Mutationhook for handleRemoveCartItems, export laoding state

Signed-off-by: Janus Reith <mail@janusreith.de>

Commit:
Remove logger - We don't want chalk tbundled on the client

Signed-off-by: Janus Reith <mail@janusreith.de>

mikemurray commented on components/MediaGallery/MediaGallery.js:
```diff
@@ -2,7 +2,8 @@ import React, { Component } from "react";
 import Grid from "@material-ui/core/Grid";
 import withStyles from "@material-ui/core/styles/withStyles";
-import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
+// import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
```
Remove commented-out code

mikemurray commented on components/MediaGalleryItem/MediaGalleryItem.js:
```diff
@@ -2,7 +2,8 @@ import React, { Component } from "react";
 import ButtonBase from "@material-ui/core/ButtonBase";
 import withStyles from "@material-ui/core/styles/withStyles";
-import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
+// import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
```
Remove commented-out code

janus-reith replied:
Regarding the commented component-library imports, they where from a point when there was no decision regarding removal and I didn't know yet if I could actually just reimplement in here, or would need to push those changes to the libary instead. Will remove them.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on components/CheckoutActions/CheckoutActions.js:
```diff
@@ -13,22 +12,24 @@ import { addTypographyStyles } from "@reactioncommerce/components/utils";
+// import TRACKING from "lib/tracking/constants";
+// import trackCheckout from "lib/tracking/trackCheckout";
+// import trackOrder from "lib/tracking/trackOrder";
+// import trackCheckoutStep from "lib/tracking/trackCheckoutStep";
```
Remove all commented-out code if it's not going to be re-enabled or used in the future.

janus-reith replied:
Ah, yes, I wanted to discuss this, but didn't investigate options yet, I agree these should definitely be adressed prior merging.
It became clear that the current approach with the tracking decorators wouldn't really be suitable for further development, and also was not really comfortable to work with.

During removal of the custom babel configuration I had not decided yet what to do with them, but did not want to remove all the references to tracking functionality upfront, in case there might be a simple way to adapt them.

Im open for suggestions here - I could just remove them from now, as we could also lookup references from the existing branches - Also, lot's of them seem obsolete and could probably better be abstracted in the hook for the respective action instead.

In the long run, I would need to research how to use react-tracking with hooks now, and maybe also evauate if it still is a good choice.
Would tracking be a priority for an initial merge?

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on components/MiniCart/MiniCart.js:
```diff
@@ -1,24 +1,23 @@
-import trackCartItems from "lib/tracking/trackCartItems";
-import track from "lib/tracking/track";
+// import trackCartItems from "lib/tracking/trackCartItems";
+// import track from "lib/tracking/track";
```
Remove commented-out code if it's not going to be used in the future.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on components/ProductDetail/ProductDetail.js:
```diff
@@ -3,8 +3,8 @@ import PropTypes from "prop-types";
-import { inject, observer } from "mobx-react";
-import track from "lib/tracking/track";
+import inject from "hocs/inject";
+// import track from "lib/tracking/track";
```
Remove commented-out code if it's not going to be used in the future.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on components/ProductGrid/ProductGrid.js:
```diff
@@ -3,8 +3,8 @@ import PropTypes from "prop-types";
-import track from "lib/tracking/track";
-import trackProductClicked from "lib/tracking/trackProductClicked";
+// import track from "lib/tracking/track";
+// import trackProductClicked from "lib/tracking/trackProductClicked";
```
Remove commented-out code if it's not going to be used in the future.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on context/RoutingContext.js:
```diff
@@ -0,0 +1,88 @@
+  }, [route, currentRoute, asPath]);
+
+  const setSearch = (search) => {
+    /*
```
Remove commented-out code

willopez replied:
ditto, all this commented code can be removed.

janus-reith replied:
Ah, yes, I wanted to discuss this, but didn't investigate options yet, I agree these should defintely be adressed prior merging.
It became clear that the current approach with the tracking decorators wouldn't really be suitable for further development, and also was not really comfortable to work with.

During removal of the custom babel configuration I had not decided yet what to do with them, but did not want to remove all the references to tracking functionality upfront, in case there might be a simple way to adapt them.

Im open for suggestions here - I could just remove them from now, as we could also lookup references from the existing branches - Also, lot's of them seem obsolete and could probably better be abstracted in the hook for the respective action instead.

In the long run, I would need to research how to use react-tracking with hooks now, and maybe also evauate if it still is a good choice.
Would tracking be a priority for an initial merge?

janus-reith replied:
Regarding the commented component-library imports, they where from a point when there was no decision regarding removal and I didn't know yet if I could actually just reimplement in here, or would need to push those changes to the libary instead. Will remove them.

janus-reith replied:
I will need to recheck the purpose of `setSearch`, I don't recall if removing is actually okay or if I might need to rewrite that instead.

willopez replied:
sounds good 👍

mikemurray commented on custom/componentsContext.js:
```diff
@@ -53,7 +54,8 @@ import MiniCartSummary from "@reactioncommerce/components/MiniCartSummary/v1";
 import ProfileImage from "@reactioncommerce/components/ProfileImage/v1";
-import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
+// import ProgressiveImage from "@reactioncommerce/components/ProgressiveImage/v1";
+import ProgressiveImage from "components/ProgressiveImage";
```
Remove commented-out imports

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on pages/[lang]/cart.js:
```diff
@@ -11,23 +11,29 @@ import withCart from "containers/cart/withCart";
-import trackCartItems from "lib/tracking/trackCartItems";
-import TRACKING from "lib/tracking/constants";
+// import trackCartItems from "lib/tracking/trackCartItems";
+// import TRACKING from "lib/tracking/constants";
```
Remove commented-out code if it’s not going to be used in the future.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray commented on pages/[lang]/index.js:
```diff
@@ -1,20 +1,22 @@
 import ProductGrid from "components/ProductGrid";
-import trackProductListViewed from "lib/tracking/trackProductListViewed";
+import Layout from "components/Layout";
+// import trackProductListViewed from "lib/tracking/trackProductListViewed";
```
Remove commented-out code if it’s not going to be used in the future.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

mikemurray reviewed (dismissed):
There's some commented out code in various files that I'd like to see addressed. Some of them can probably be removed like the commented out imports from the `component-library`.

I'm not sure why the tack event stuff is commented out, or if that code will be revisited in the future. I feel like you may address it in a future PR, but I wanted to be sure.

willopez commented on package.json:
```diff
@@ -35,73 +35,53 @@
     "keymirror": "~0.1.1",
     "lodash": "~4.17.11",
     "mdi-material-ui": "~5.9.0",
     "mobx": "~4.9.2",
```
Remove `mobx` and `mobx-react` as they are no longer needed. Also, remove related documentation regarding on how to work with mobx store.s

janus-reith replied:
Removed Mobx-related docs in https://github.com/reactioncommerce/example-storefront/pull/667/commits/02925958b57284eb5553cde5231bc34bcf729e5d, removed Mobx in https://github.com/reactioncommerce/example-storefront/pull/667/commits/58af0d8a5bd00a2baa5d6e7ecd11dfa8f6850ced

willopez commented on package.json:
```diff
@@ -35,73 +35,53 @@
-    "apollo-link-error": "~1.1.7",
-    "apollo-link-http": "~1.5.11",
-    "apollo-link-ws": "~1.0.14",
+    "apollo-client": "^2.6.8",
```
are both Apollo client packages needed? `3.0.0-beta.41` and `2.6.8`

janus-reith replied:
I believe apollo-client is obsolete, will verify and remove.

janus-reith replied:
https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

janus-reith replied:
`apollo-client` is required by `@apollo/react-components`, which provides the `Query`component used in components/ProfileContainer and containers/catalog/withCatalogItems.

Once these two are migrated to hooks, we can remove the old `ApolloProvider` in `withApollo` and remove both `@apollo/react-components` and `apollo-client`

willopez replied:
👍

willopez commented on components/ProductDetail/ProductDetail.js:
```diff
@@ -168,11 +155,12 @@ class ProductDetail extends Component {
         // therefore, a CART_VIEWED event is published.
         // debugger // eslint-disable-line
-        this.trackCartItems({ cartItems: items, cartId: cart._id, action: CART_VIEWED }); // eslint-disable-line camelcase
+        // this.trackCartItems({ cartItems: items, cartId: cart._id, action: CART_VIEWED }); // eslint-disable-line camelcase
```
We can either remove all the tracking code, which would not be my preference since tracking user actions is an important aspect of any store, or refactor the tracking as well. What is your intention with the tracking, were you planning or removing it or refactoring?

janus-reith replied:
Ah, yes, I wanted to discuss this, but didn't investigate options yet, I agree these should definitely be adressed prior merging.
It became clear that the current approach with the tracking decorators wouldn't really be suitable for further development, and also was not really comfortable to work with.

During removal of the custom babel configuration I had not decided yet what to do with them, but did not want to remove all the references to tracking functionality upfront, in case there might be a simple way to adapt them.

Im open for suggestions here - I could just remove them for now, as we could also lookup references from the existing branches - Also, lot's of them seem obsolete and could probably better be abstracted in the hook for the respective action instead.

In the longer run, I would need to research how to use react-tracking with hooks now, and maybe also evaluate if it still is a good choice.
Would tracking be a priority for an initial merge?

willopez replied:
Let's remove the current tracking implementation and create an issue to re-implement using a modern hooks based approach.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

willopez commented on custom/analytics/provider.example.js:
```diff
@@ -18,9 +16,7 @@ export function dispatch(data) {
-  // An API key variable may be set in `publicRuntimeConfig` object of the next.config.js
-  const { publicRuntimeConfig } = getConfig();
-  const { apiKey } = publicRuntimeConfig; // eslint-disable-line no-unused-vars
+  // An API key variable may be set in `env` object of the next.config.js
```
if tracking will be removed, then this needs to be removed to or refactored to work.

janus-reith replied:
Removed in https://github.com/reactioncommerce/example-storefront/pull/667/commits/f99f9adf797a20148632641ef9413b915e711c56

willopez commented on custom/analytics/segment.js:
```diff
@@ -23,14 +22,12 @@ export function dispatch(data) {
-
   const opts = {
-    apiKey: segmentAnalytics.writeKey,
+    apiKey: process.env.SEGMENT_ANALYTICS_WRITE_KEY,
```
same comment as all the other tracking code.

willopez commented on package.json:
```diff
@@ -13,18 +12,19 @@
     "dev": "echo 'Note: dev script has been renamed to start:dev' && yarn run start:dev",
-    "build": "yarn run build-clean && NODE_ENV=production IS_BUILDING_NEXTJS=1 next build src",
+    "build": "yarn run build-clean && NODE_ENV=production IS_BUILDING_NEXTJS=1 next build",
     "build-clean": "rimraf build/app",
```
Seems like `build-clean` command is no longer needed, can be removed from this line and the one above.

janus-reith replied:
Correct, there is no build folder anymore.
I could add a clean script which remove the .next folder instead, thereby removing builds and cached pages.

willopez replied:
mmm, if it's necessary, then let's add it, otherwise might not be needed anymore. Not sure, it's your call.

janus-reith replied:
Just removed it for now in https://github.com/reactioncommerce/example-storefront/pull/667/commits/c02077a65446dd23c09be7ec925879af8e7401a0.

willopez reviewed (changes requested):
Fantastic work! This makes the storefront much more useful right out of the box. Thank you again for this great effort. Left a couple of comments/questions and some small change requests.

willopez commented:
A couple of more issues I fount: 

1. Building the production images fails due to the movement of the source files. Please update the `Dockerfile` accordinly.

2. When building the production build by executing `yarn build` I see the following errors:
```
Error occurred prerendering page "/de/product/-". Read more: https://err.sh/next.js/prerender-error:
ReferenceError: ErrorPage is not defined
```
```
error-fetching-graphql Error: Tag not found: {"response":{"errors":[{"message":"Tag not found","locations":[{"line":3,"column":5}],"path":["tag"],"extensions":{"code":"NOT_FOUND","exception":{"eventData":{},"error":"not-found","isClientSafe":true,"reason":"Tag not found","details":{},"stacktrace":["ReactionError: Tag not found","    at Object.tag (file:///usr/local/src/app/src/core-services/tags/queries/tag.js:26:11)","    at runMicrotasks (<anonymous>)","    at processTicksAndRejections (internal/process/task_queues.js:94:5)"]}}}],"data":{"tag":null},"status":200},"request":{"query":"\nquery tagQuery($shopId: ID!, $slugOrId: String!) {\n    tag(shopId: $shopId, slugOrId: $slugOrId) {\n      ...TagInfo\n    }\n}\nfragment TagInfo on Tag {\n    _id\n    position\n    name\n    slug\n    isTopLevel\n    subTagIds\n    heroMediaUrl\n    metafields {\n      key\n      namespace\n      scope\n      value\n    }\n    displayTitle\n}  \n","variables":{"shopId":"cmVhY3Rpb24vc2hvcDpOUUxhdnljYmdtUHhBWDh6cw==","slugOrId":"-"}}}
    at GraphQLClient.<anonymous> (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:116:35)
    at step (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:40:23)
    at Object.next (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:21:53)
    at fulfilled (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:12:58)
    at processTicksAndRejections (internal/process/task_queues.js:94:5) {
  response: { errors: [ [Object] ], data: { tag: null }, status: 200 },
```

3. In the browser console, I see a the following warnings and one error: 
```
Warning: React.createFactory() is deprecated and will be removed in a future major release
```
That warning is coming from the component library, we could just create an issue in that repo and resolve it there. Or just don't use the `QuantityInput` component. This could be done in another PR, so that this one does not keep getting bigger.

```
Unknown event handler property `onItemClick`. It will be ignored.
    in div (created by CatalogGrid__GridItem)
    in CatalogGrid__GridItem (created by ContainerQuery)
```
Error above is due to the tracking code, we could refactor this or just remove it.

Overall, great work! Looking forward to merging this.

willopez commented:
Also found the following security vulnerabilities: 
```
Issues to fix by upgrading:

  Upgrade next@9.3.0 to next@9.3.2 to fix
  ✗ Prototype Pollution [Medium Severity][https://snyk.io/vuln/SNYK-JS-MINIMIST-559764] in minimist@0.0.8
    introduced by next@9.3.0 > mkdirp@0.5.1 > minimist@0.0.8 and 26 other path(s)
  ✗ Path Traversal (new) [Medium Severity][https://snyk.io/vuln/SNYK-JS-NEXT-561584] in next@9.3.0

Issues with no direct upgrade or patch:
  ✗ Timing Attack [Medium Severity][https://snyk.io/vuln/SNYK-JS-ELLIPTIC-511941] in elliptic@6.5.1
    introduced by next@9.3.0 > webpack@4.42.0 > node-libs-browser@2.2.1 > crypto-browserify@3.12.0 > browserify-sign@4.0.4 > elliptic@6.5.1 and 1 other path(s)
  This issue was fixed in versions: 6.5.2
  ✗ Information Exposure [Low Severity][https://snyk.io/vuln/SNYK-JS-KINDOF-537849] in kind-of@6.0.2
    introduced by next@9.3.0 > fork-ts-checker-webpack-plugin@3.1.1 > micromatch@3.1.10 > kind-of@6.0.2 and 186 other path(s)
  This issue was fixed in versions: 6.0.3
```

Looks like upgrading to the latest version (`9.3.4`) of Next.js should fix them.

janus-reith commented:
> Also found the following security vulnerabilities:
> 
> ```
> Issues to fix by upgrading:
> 
>   Upgrade next@9.3.0 to next@9.3.2 to fix
>   ✗ Prototype Pollution [Medium Severity][https://snyk.io/vuln/SNYK-JS-MINIMIST-559764] in minimist@0.0.8
>     introduced by next@9.3.0 > mkdirp@0.5.1 > minimist@0.0.8 and 26 other path(s)
>   ✗ Path Traversal (new) [Medium Severity][https://snyk.io/vuln/SNYK-JS-NEXT-561584] in next@9.3.0
> 
> Issues with no direct upgrade or patch:
>   ✗ Timing Attack [Medium Severity][https://snyk.io/vuln/SNYK-JS-ELLIPTIC-511941] in elliptic@6.5.1
>     introduced by next@9.3.0 > webpack@4.42.0 > node-libs-browser@2.2.1 > crypto-browserify@3.12.0 > browserify-sign@4.0.4 > elliptic@6.5.1 and 1 other path(s)
>   This issue was fixed in versions: 6.5.2
>   ✗ Information Exposure [Low Severity][https://snyk.io/vuln/SNYK-JS-KINDOF-537849] in kind-of@6.0.2
>     introduced by next@9.3.0 > fork-ts-checker-webpack-plugin@3.1.1 > micromatch@3.1.10 > kind-of@6.0.2 and 186 other path(s)
>   This issue was fixed in versions: 6.0.3
> ```
> 
> Looks like upgrading to the latest version (`9.3.4`) of Next.js should fix them.

Yes, will update nextjs, altough Im still waiting for a stable v9.3.5 release and was unsure wether we could just use a canary version - v9.3.5-canary.8 fixed some cpu/memory leaking issue and increases performance for fallback creation a lot.

janus-reith commented:
> A couple of more issues I fount:
> 
> 1. Building the production images fails due to the movement of the source files. Please update the `Dockerfile` accordinly.
> 2. When building the production build by executing `yarn build` I see the following errors:
> 
> ```
> Error occurred prerendering page "/de/product/-". Read more: https://err.sh/next.js/prerender-error:
> ReferenceError: ErrorPage is not defined
> ```
> 
> ```
> error-fetching-graphql Error: Tag not found: {"response":{"errors":[{"message":"Tag not found","locations":[{"line":3,"column":5}],"path":["tag"],"extensions":{"code":"NOT_FOUND","exception":{"eventData":{},"error":"not-found","isClientSafe":true,"reason":"Tag not found","details":{},"stacktrace":["ReactionError: Tag not found","    at Object.tag (file:///usr/local/src/app/src/core-services/tags/queries/tag.js:26:11)","    at runMicrotasks (<anonymous>)","    at processTicksAndRejections (internal/process/task_queues.js:94:5)"]}}}],"data":{"tag":null},"status":200},"request":{"query":"\nquery tagQuery($shopId: ID!, $slugOrId: String!) {\n    tag(shopId: $shopId, slugOrId: $slugOrId) {\n      ...TagInfo\n    }\n}\nfragment TagInfo on Tag {\n    _id\n    position\n    name\n    slug\n    isTopLevel\n    subTagIds\n    heroMediaUrl\n    metafields {\n      key\n      namespace\n      scope\n      value\n    }\n    displayTitle\n}  \n","variables":{"shopId":"cmVhY3Rpb24vc2hvcDpOUUxhdnljYmdtUHhBWDh6cw==","slugOrId":"-"}}}
>     at GraphQLClient.<anonymous> (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:116:35)
>     at step (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:40:23)
>     at Object.next (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:21:53)
>     at fulfilled (/usr/local/src/app/node_modules/graphql-request/dist/src/index.js:12:58)
>     at processTicksAndRejections (internal/process/task_queues.js:94:5) {
>   response: { errors: [ [Object] ], data: { tag: null }, status: 200 },
> ```
> 
> 1. In the browser console, I see a the following warnings and one error:
> 
> ```
> Warning: React.createFactory() is deprecated and will be removed in a future major release
> ```
> 
> That warning is coming from the component library, we could just create an issue in that repo and resolve it there. Or just don't use the `QuantityInput` component. This could be done in another PR, so that this one does not keep getting bigger.
> 
> ```
> Unknown event handler property `onItemClick`. It will be ignored.
>     in div (created by CatalogGrid__GridItem)
>     in CatalogGrid__GridItem (created by ContainerQuery)
> ```
> 
> Error above is due to the tracking code, we could refactor this or just remove it.
> 
> Overall, great work! Looking forward to merging this.

That looks like some little oversights, will try to fix both tomorrow.
Regarding 3, yes I noticed that aswell and didn't take further action as it originates from the component library and is just a warning. But as it really looks weird and might disturb developer ux, I'm happy to quickly rewrite it in the storefront with just MUI.

willopez commented:
We can update to the current latest version of next now, and when the new one come out, it can be updated and it should be a quick new PR.

janus-reith committed:
Upgrade nextjs to 9.3.5

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Remove commented-out code and obsolete tracking

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith commented:
> We can update to the current latest version of next now, and when the new one come out, it can be updated and it should be a quick new PR.

Upgraded to 9.3.5 in https://github.com/janus-reith/example-storefront/commit/c06f84113d5f8ee4ed594b01b4d0e3b350629fb2

janus-reith committed:
Remove apollo-client

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Remove MobX-related documentation

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Fix outdated ApolloLink import

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Add back apollo-client, as @apollo/react-components used in ProfileContainer and withCatalogItems use it TODO: Migrate both and remove

Signed-off-by: Janus Reith <mail@janusreith.de>

Revert "Add back apollo-client, as @apollo/react-components used in ProfileContainer and withCatalogItems use it TODO: Migrate both and remove"

This reverts commit 8038c5e6e57508095eb2ae9f6da12bce6f839691.

Add back apollo-client, as @apollo/react-components used in ProfileContainer and withCatalogItems use it TODO: Migrate both and remove

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Remove ErrorPage and use simple Typography text for now. TODO: replace with loading Skeleton

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Remove build-clean script

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith commented:
Some things are a bit hard to keep track, I made a small list of things left TBD:

- [ ] Recheck `setSearch` and remove or replace it in `context/RoutingContext.js`
- [ ] Migrate `components/ProfileContainer` and `containers/catalog/withCatalogItems` to hooks and remove the Provider in `lib/apollo/withApollo` to be able to remove `@apollo/react-components` and `apollo-client`
- [x] Check and fix the Dockerfile for production builds, make sure it works for default setups when the repo has no override for dev mode.
- [ ] Remove or replace `custom/analytics/provider.example.js` and `custom/analytics/segment.js` depending on how we implement this.
- [ ] Resolve the `React.createFactory()` warning by moving that component off the libary and into a new simple MUI component.
- [ ] Avoid styling issues resulting from library version mismatches by removing all references to the component library and move things over to the storefront repo.
- [ ] Find out what causes `error-fetching-graphql Error: Tag not found`. This occurs during the fetch for static data called in `getStaticProps`. We could consider adding a log to console for each of the functions I created in `staticUtils`, to better determine what slug/_id/name was used in a call, also later durng runtime if a page is revalidated or created during fallback.

Soon but not immediately:
- [ ] Add new Docs, primarily describing the patters how the hooks and Providers for React Context and static data fetching works.

Maybe later:
- [ ] "-" slug in "/de/product/-" which caused `ErrorPage is not defined`. I could simply resolve the error by removing the `ErrorPage` and using a simple Typography component, at some point I want to move this to MUI skeleton loaders. But the reason that the error page would actually show is the way fallback works: I need to provide a list of slugs to be used for `getStaticPaths` which must not be empty, the rest would be fallback-created during runtime. Therefore I added "-" as the slug for the product page created during build as we don't know the product pages. We could instead do something like fetching the first 10 products during build, but that would not work for cases where we build a storefront and have no products, and we would also need to use a stub in that case, so as there is no "perfect" solution for now we can just leave it this way.
- [ ] Expanding upon `custom/analytics`: We could still keep this pattern, as it is not directly related to react-tracking and could also work for other scripts. The general pattern used in there to add scripts to `_document.js` could use another discussion though, as for this to be actually helpful IMHO it needs to be more sophisticated, like allowing both inline or linked scripts or `async/defer` defined per script.

janus-reith commented:
DCO failed again after one of my commits was a revert, and I have no idea why.
I tried to use `--amend` as DCO suggest, and also rebased everything after it still failed, and Github shows me that https://github.com/reactioncommerce/example-storefront/pull/667/commits/0893b04b2da7ff6d4559aba01c362d0482ff55aa is verified properly - DCO still fails stating that the sign-off is missing.

janus-reith committed:
RFix next _document.js according to https://github.com/mui-org/material-ui/pull/20463

Signed-off-by: Janus Reith <mail@janusreith.de>

willopez commented:
looking at commit  https://github.com/reactioncommerce/example-storefront/pull/667/commits/0893b04b2da7ff6d4559aba01c362d0482ff55aa I don't see the `Signed-off-by:` signature, so DOC is correctly reporting that is not signed off. The instructions DOC provides on how to resolve unsigned commits only apply if the unsigned commit was the last one pushed, so it will not work for older commits :( unfortunately. I believe you can go back in git history and squash(merged) the unsigned commit with one that is signed and that should do it. Here is a guide: https://www.atlassian.com/git/tutorials/rewriting-history

janus-reith commented:
@willopez Thanks, will try again.
If I click that commit, Github shows it as verified in a green box.
As signing just the last commit as described by DCO expectedly did not work, I already tried rebasing everything twice, starting from the first commit I made in this branch.
Looking at the commit history here, everything seems verified:
https://github.com/janus-reith/example-storefront/commits/storefront-v2
But you're right, that "Signed-off-by" is missing from that Revert commit.

I might rebase again and this time remove  both the original commit and its revert in the process, still I'd like to figure out what's going on.

janus-reith committed:
Upgrade nextjs to v9.4.1

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Fix the docker build and provide a default env

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith committed:
Remove commented out code

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith commented:
I just fixed the build of the docker container container for production builds, here are some things to note:

- The API needs to be accessible during build

- I created .env.prod which links to the default setup and excluded that in .gitingore from being ignored. You can pass a build arument NEXTJS_DOTENV, which will then be used after the dot, e,g, NEXTJS_DOTENV=abc would result in a file .env.abc being picked up. prod is the default fallback value, but I'm open to alternative approaches. I mainly used this myself to be able to generate different deployments from the same source code.

- During build, the graphl utility I made in staticUtils/graphQLRequest.js checks for the IS_BUILDING_NEXTJS  env var, and if true defaults to fetching from the external api URL as it is assumed that the build container won't share a docker network with the api container

- Currently, when running on the same host, you need to pass `--network=host` so the build container is able to locate the ap running at localhost:3000, so the call might be: `docker build --network=host -t reactioncommerce/example-storefront:3.0.0 .`
There might be other ways to resolve the route to the API from inside the container, but I found none yet. Also, that concern might be irrelevant when running in a CI and the API is running on some external host anyway.

I expect some adjustment to be necesary so your CI will build the generic image with that, but for now the configuration will work out to build the image locally.
One issue that might come up: During build on a CI, `process.env.EXTERNAL_GRAPHQL_URL` will be used to fetch data for the static pages, and from then will be inlined to be used by the `withApollo` HOC clientSide, in other words, you wont be able to just spin up the image and pass your own value for that. This is a security measure by nextjs to avoid leaking sensitive env vars to the client.
`process.env.INTERNAL_GRAPHQL_URL` on the other hand will only be used internally when static fallback pages are created during runtime.
I'm currently considering to introduce a third variable, BUILD_GRAPHQL_URL, to allow for a scenario like when the default non-dev example image is built on a CI:
BUILD_GRAPHQL_URL could be something external like api.reactioncommerce.com, while EXTERNAL_GRAPHQL_URL might still point to localhost and INTERNAL_GRAPHQL_URL to the internal docker network.

In general, some of the things regarding env vars tend to be tricky, as nextjs inlines some vars if defined in next.config.js while some others stay dynamic and developers will need to take a close look at `config.js` and `next.config.js` when customizing to see how things behave in client and server contexts.

janus-reith committed:
Introduce new BUILD_GRAPHQL_URL env var

Signed-off-by: Janus Reith <mail@janusreith.de>

janus-reith commented:
I just included `BUILD_GRAPHQL_URL`, as there are no real alternatives for decoupled builds that should run locally and it also might be easier this way to comprehend what is used where.

I didn't bother about that until now, as usually I would either use the dev image locally, or have the production build running against the api endpoint it would also be used with later, and I don't expect many workflows to actually make use of deviating `BUILD_GRAPHQL_URL`/`EXTERNAL_GRAPHQL_URL` vars except for the demo image, still it is reasonable to be present.

Maybe @focusaurus could chime in on what to do best with the fragmentation of .env files we have now? There now is `.env.example` which only has the purpouse of being copied over to `.env` which will be used for dynamic parts like the api route, and third there is .env.prod which currently is expected to be picked up during build. But ideally I would want to deal with this in a separate  follow up PR, considering that his might also introduce changes to the platform dev script.

willopez commented:
@janus-reith The production builds are working now, this is a good start. If issues arise, we can iterate on the process. Thanks!

willopez reviewed (approved):
LGTM. There are known issues/improvements that will be resolved in subsequent PRs, I'll create tickets for those issues/improvements.

willopez commented:
@janus-reith Thanks again for this contribution, I have created issues for the known issues that can be resolved in subsequent PRs.
